### PR TITLE
Implement security for rmw-cyclone

### PR DESCRIPTION
**Pull request by SidFaber**
Initial implementation; still needs extensive testing.

### CODE REVIEW

File: rmw_cyclonedds_cpp/src/rmw_node.cpp [hunk 1]
```diff
@@ -688,10 +765,56 @@ extern "C" rmw_node_t * rmw_create_node(
 #endif
 
   dds_qos_t * qos = dds_create_qos();
+  if (!qos) {
```
Comment (SidFaber) [inline rmw_cyclonedds_cpp/src/rmw_node.cpp]: if qos == nullptr

File: rmw_cyclonedds_cpp/src/rmw_node.cpp [hunk 2]
```diff
@@ -666,7 +737,13 @@ extern "C" rmw_node_t * rmw_create_node(
   static_cast<void>(domain_id);
   const dds_domainid_t did = DDS_DOMAIN_DEFAULT;
 #endif
-  (void) security_options;
+
+  if (!security_options) {
```
Comment (SidFaber) [inline rmw_cyclonedds_cpp/src/rmw_node.cpp]: if security_options == nullptr?

File: rmw_cyclonedds_cpp/src/rmw_node.cpp [hunk 3]
```diff
@@ -688,10 +765,56 @@ extern "C" rmw_node_t * rmw_create_node(
...
   dds_qos_t * qos = dds_create_qos();
+  if (!qos) {
+      RCUTILS_LOG_ERROR_NAMED(
+        "rmw_cyclonedds_cpp", "rmw_create_node: Unable to create qos");
+      return nullptr;
+  }
   std::string user_data = get_node_user_data(name, namespace_);
   dds_qset_userdata(qos, user_data.c_str(), user_data.size());
+
+//add in a #IF on whether security is compiled in?
```
Comment (SidFaber) [inline rmw_cyclonedds_cpp/src/rmw_node.cpp]: check with the rest of the cyclone codebased

File: rmw_cyclonedds_cpp/src/rmw_node.cpp [hunk 4]
```diff
@@ -688,10 +765,56 @@ extern "C" rmw_node_t * rmw_create_node(
...
+  if (security_options->enforce_security) {
+    char * security_file_paths [SFT_MAX_COUNT];
+    if (!(security_options->security_root_path)) {
+      RCUTILS_LOG_ERROR_NAMED(
+        "rmw_cyclonedds_cpp", "rmw_create_node: no securty path.");
+      return nullptr;
+    }
+    if (not (get_security_file_URIs(security_file_paths, security_options->security_root_path))) {
```
Comment (SidFaber) [inline rmw_cyclonedds_cpp/src/rmw_node.cpp]:
```suggestion
    if (! (get_security_file_URIs(security_file_paths, security_options->security_root_path))) {
```

File: rmw_cyclonedds_cpp/src/rmw_node.cpp [hunk 5]
```diff
@@ -642,6 +644,75 @@ static std::string get_node_user_data(const char * node_name, const char * node_
...
+        URIs as well.  */
+    security_file_URIs[i] = rcutils_format_string(allocator, "file:%s", file_path);
+    allocator.deallocate(file_path, allocator.state);
+    if (!security_file_URIs[i]) {
+      finalize_security_file_URIs(security_file_URIs);
+      return false;
+    }
+  }
+  return true;
+}
```
Comment (kyrofa) [inline rmw_cyclonedds_cpp/src/rmw_node.cpp]: I think what you've done here is far more clear than relying on raw indices into a list. However, I don't think it goes far enough; we can still simplify it and improve readability by no longer using a list at all, but a class. Can we do that? Before I continue I will point out that:

1. I haven't tested this code at _all_ and I doubt it'll compile as-is, but hopefully it gives you an idea
2. I haven't looked at the code style here at all

First of all, I suggest defining the file names [up here](https://github.com/ubuntu-robotics/rmw_cyclonedds/pull/1/files#diff-d03c99cafce317e1a93de46e787e0235R113) along with other similar variables. They should probably be an in anonymous namespace, but they aren't, so follow their pattern, e.g.:

```c++
const char * const eclipse_cyclonedds_identity_ca_cert_file_name = "identity_ca.cert.pem";
const char * const eclipse_cyclonedds_cert_file_name = "identity_ca.cert.pem";
// ...
```

Then we have :

```c++
class SecurityFilesFinder {
  public:
    explicit SecurityFiles(const char *node_secure_root, rcutils_allocator_t allocator = rcutils_get_default_allocator()) :
      m_allocator(allocator),
      m_identity_ca_cert(rcutils_join_path(node_secure_root, eclipse_cyclonedds_identity_ca_cert_file_name, m_allocator),
      m_cert(rcutils_join_path(node_secure_root, eclipse_cyclonedds_cert_file_name, m_allocator)
    {
    }

    ~SecurityFiles()
    {
      m_allocator.deallocate(m_identity_ca_cert);
      m_allocator.deallocate(m_cert);
      // ...
    }

    bool check_files_are_readable() const
    {
      return (
        (m_identity_ca_cert != nullptr) && rcutils_is_readable(m_identity_ca_cert) &&
        (m_cert != nullptr) && rcutils_is_readable(m_cert) &&
        // ...
      );
    }

  const char * const identity_ca_cert() const
  {
    return m_identity_ca_cert;
  }

  const char * const cert() const
  {
    return m_cert;
  }

  private:
    char *m_identity_ca_cert;
    char *m_cert;
    // ...
};
```

That nicely scopes the deallocation so you don't need to deal with it explicitly. Then we can directly use this class below:

```c++
  if (security_options->enforce_security) {
    if (!(security_options->security_root_path)) {
      RCUTILS_LOG_ERROR_NAMED(
        "rmw_cyclonedds_cpp", "rmw_create_node: no securty path.");
      return nullptr;
    }

    SecurityFilesFinder security_files_finder(security_options->security_root_path);

    if (!security_files_finder.check_files_are_readable()) {
      RCUTILS_LOG_ERROR_NAMED(
        "rmw_cyclonedds_cpp", "rmw_create_node: unable to locate security files.");
      return nullptr;
    }

    dds_qset_prop(qos, "dds.sec.auth.identity_ca", security_files_finder.identity_ca_cert());
    dds_qset_prop(qos, "dds.sec.auth.identity_certificate", security_files_finder.cert());
    // ...
```

What do you think?

File: rmw_cyclonedds_cpp/src/rmw_node.cpp (hunk 2 above)
Comment (kyrofa) [reply rmw_cyclonedds_cpp/src/rmw_node.cpp]: Yeah, let's do that.

File: rmw_cyclonedds_cpp/src/rmw_node.cpp [hunk 6]
```diff
@@ -688,10 +765,56 @@ extern "C" rmw_node_t * rmw_create_node(
...
+      RCUTILS_LOG_ERROR_NAMED(
+        "rmw_cyclonedds_cpp", "rmw_create_node: Unable to create qos");
+      return nullptr;
+  }
   std::string user_data = get_node_user_data(name, namespace_);
   dds_qset_userdata(qos, user_data.c_str(), user_data.size());
+
+//add in a #IF on whether security is compiled in?
+  if (security_options->enforce_security) {
+    char * security_file_paths [SFT_MAX_COUNT];
```
Comment (kyrofa) [inline rmw_cyclonedds_cpp/src/rmw_node.cpp]: See suggestion above. Also consider extracting to a standalone function.

**Pull request closed without merging**